odbc/src/api/functions.rs

```diff
@@ -1279,72 +1274,183 @@ pub unsafe extern "C" fn SQLFetch(statement_handle: HStmt) -> SqlReturn {
+        let row_status_buffer: *mut USmallInt =
+            if !stmt.attributes.read().unwrap().row_status_ptr.is_null() {
+                (stmt.attributes.read().unwrap().row_status_ptr as ULen
+                    + (index * size_of::<u16>())) as *mut USmallInt
+            } else {
+                null_mut()
+            };
+
+        if let Ok((has_next, mut warnings_opt)) = move_to_next_result {
+            warnings_opt.iter().for_each(|warning| {
+                add_diag_with_function!(
+                    MongoHandleRef::from(statement_handle),
+                    ODBCError::GeneralWarning(warning.to_string()),
+                    function_name.to_string()
                 );
+            });
+
+            if !has_next {
+                // There are no more rows, so set the rest of Row Status Array buffers to SQL_ROW_NOROW if the rowset end has not been reached.
+                if !row_status_buffer.is_null() {
+                    for row_status_index in index..rowset_size {
+                        let row_status_buffer = (stmt.attributes.read().unwrap().row_status_ptr
+                            as ULen
+                            + (row_status_index * size_of::<u16>()))
+                            as *mut USmallInt;
+                        *row_status_buffer = definitions::SQL_ROW_NOROW;
+                    }
+                }
+
+                // break here to prevent NO_DATA from being returned if the rowset had at least one row in it.
+                if has_fetched_at_least_one_row {
+                    break;
+                }
+
+                // if the cursor has already reached the end of the result set when SQLFetch is called, NO_DATA is returned.
+                // Additionally, rows_fetch_ptr is set to 0.
+                return SqlReturn::NO_DATA;
+            }
+
+            has_fetched_at_least_one_row = true;
+
+            if !row_status_buffer.is_null() {
+                *row_status_buffer = if warnings_opt.is_empty() {
+                    definitions::SQL_ROW_SUCCESS
+                } else {
+                    definitions::SQL_ROW_SUCCESS_WITH_INFO
+                };
+            }
 
-                match sql_return {
-                    SqlReturn::ERROR => return SqlReturn::ERROR,
-                    SqlReturn::SUCCESS_WITH_INFO => {
-                        success_with_info_encountered = true;
+            // keep track of all warnings that occur throughout the entire function.
+            global_warnings_opt.append(&mut warnings_opt);
+
+            if has_rows_fetched_buffer {
+                *stmt.attributes.write().unwrap().rows_fetched_ptr += 1;
+            }
+
+            *stmt.var_data_cache.write().unwrap() = Some(HashMap::new());
+
+            // If there are bound columns, then copy data from the result set into the bound buffers.
+            if let Some(bound_cols) = stmt.bound_cols.read().unwrap().as_ref() {
```

Comment:
Since `sql_fetch_helper` is getting pretty long, I could put the column binding part in a separate helper function if that would be preferable.

Reply:
I would be in favor of this with readability in mind!